cpp/src/explainer/tree_shap.cu

```diff
@@ -0,0 +1,326 @@
+      if (!tree.IsLeaf(i)) {
+        parent_id[tree.LeftChild(i)] = i;
+        parent_id[tree.RightChild(i)] = i;
+      }
+    }
+
+    // Find leaf nodes
+    // Work backwards from leaf to root, order does not matter
+    // It's also possible to work from root to leaf
+    for (int i = 0; i < tree.num_nodes; i++) {
+      if (tree.IsLeaf(i)) {
+        float v = static_cast<float>(tree.LeafValue(i));
+        int child_idx = i;
+        int parent_idx = parent_id[child_idx];
+        const auto inf = std::numeric_limits<ThresholdType>::infinity();
+        while (parent_idx != -1) {
+          double zero_fraction = 1.0;
+          bool has_count_info = false;
+          if (tree.HasSumHess(parent_idx) && tree.HasSumHess(child_idx)) {
+            zero_fraction = static_cast<double>(
+                tree.SumHess(child_idx) / tree.SumHess(parent_idx));
+            has_count_info = true;
+          }
+          if (tree.HasDataCount(parent_idx) && tree.HasDataCount(child_idx)) {
+            zero_fraction = static_cast<double>(
+                tree.DataCount(child_idx)) / tree.DataCount(parent_idx);
+            has_count_info = true;
+          }
+          if (!has_count_info) {
+            RAFT_FAIL("Lacking sufficient info");
+          }
+          // Encode the range of feature values that flow down this path
+          bool is_left_path = tree.LeftChild(parent_idx) == child_idx;
+          if (tree.SplitType(parent_idx) == tl::SplitFeatureType::kCategorical) {
+            RAFT_FAIL("For now only trees with numerical splits are supported. "
+                      "Trees with categorical splits are not supported yet.");
+          }
+          ThresholdType lower_bound =
+            is_left_path ? -inf : tree.Threshold(parent_idx);
+          ThresholdType upper_bound =
+            is_left_path ? tree.Threshold(parent_idx) : inf;
+          int group_id = tree_idx % num_groups;
+          paths->paths.push_back(
+              gpu_treeshap::PathElement<MySplitCondition<ThresholdType>>{
+                path_idx,
+                tree.SplitIndex(parent_idx),
+                group_id,
+                {lower_bound, upper_bound},
+                zero_fraction,
+                v});
+          child_idx = parent_idx;
+          parent_idx = parent_id[child_idx];
+        }
+        // Root node has feature -1
+        {
+          int group_id = tree_idx % num_groups;
+          paths->paths.push_back(
+              gpu_treeshap::PathElement<MySplitCondition<ThresholdType>>{
+                path_idx, -1, group_id, {-inf, inf}, 1.0, v});
+          path_idx++;
+        }
+      }
+    }
+    tree_idx++;
+  }
+  paths->global_bias = model.param.global_bias;
+  paths->task_type = model.task_type;
+  paths->task_param = model.task_param;
+  paths->average_tree_output = model.average_tree_output;
+  paths->num_tree = static_cast<int>(model.trees.size());
+
+  return static_cast<ExtractedPathHandle>(path_container.release());
```

Comment:
Why not continue to work with this as a `unique_ptr`?

Reply:
I need to implement a type erasure so that the Cython code only sees an opaque `void*` handle. The reason is that you don't know whether a given Treelite object is using a 32-bit or 64-bit float for thresholds, so you can't specify the precise type `ExtractedPath<T>` in advance.